File: internal/secrets/seal/sas_test.go
package seal

import (
	"regexp"
	"testing"
)

// sasFmt matches the "xxxx-xxxx-xxxx" 3×4 base-36 output format.
var sasFmt = regexp.MustCompile(`^[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}$`)

func TestDeriveSASFormat(t *testing.T) {
	genesis := []byte("genesis-hash-test")
	head := []byte("head-hash-test")
	x25519 := []byte("new-x25519-pub-32bytes-padded-xx")
	sign := []byte("new-sign-pub-65bytes-padded-xxxx")

	got, err := DeriveSAS(genesis, head, x25519, sign)
	if err != nil {
		t.Fatalf("DeriveSAS: %v", err)
	}
	if !sasFmt.MatchString(got) {
		t.Fatalf("SAS format wrong: %q (want xxxx-xxxx-xxxx base-36)", got)
	}
	t.Logf("SAS: %s", got)
}

func TestDeriveSASStability(t *testing.T) {
	genesis := []byte("genesis-hash-test")
	head := []byte("head-hash-test")
	x25519 := []byte("new-x25519-pub-32bytes-padded-xx")
	sign := []byte("new-sign-pub-65bytes-padded-xxxx")

	a, err := DeriveSAS(genesis, head, x25519, sign)
	if err != nil {
		t.Fatalf("first DeriveSAS: %v", err)
	}
	b, err := DeriveSAS(genesis, head, x25519, sign)
	if err != nil {
		t.Fatalf("second DeriveSAS: %v", err)
	}
	if a != b {
		t.Fatalf("DeriveSAS not stable: %q vs %q", a, b)
	}
}

// TestDeriveSASMITMSubstitutedPubkey is the spec §4 named negative: a MITM that
// substitutes the new device's x25519 public key MUST produce a different SAS,
// so the human comparison catches the attack (spec §2 [WM4]).  If the SAS were
// the same for a substituted key, an active attacker could swap the pubkey in
// transit and enroll their own device undetected.
func TestDeriveSASMITMSubstitutedPubkey(t *testing.T) {
	genesis := []byte("genesis-hash-test")
	head := []byte("head-hash-test")
	legitX25519 := []byte("legit-x25519-pub-32bytes-paddxxx")
	sign := []byte("new-sign-pub-65bytes-padded-xxxx")

	legitimate, err := DeriveSAS(genesis, head, legitX25519, sign)
	if err != nil {
		t.Fatalf("DeriveSAS legitimate: %v", err)
	}

	// Attacker substitutes a different x25519 pubkey.
	mitmX25519 := []byte("mitm--x25519-pub-32bytes-paddxxx")
	mitmSAS, err := DeriveSAS(genesis, head, mitmX25519, sign)
	if err != nil {
		t.Fatalf("DeriveSAS MITM: %v", err)
	}
	if legitimate == mitmSAS {
		t.Fatalf("MITM failure: substituted x25519 pubkey produced identical SAS %q — "+
			"a MITM'd link would pass the human comparison undetected", legitimate)
	}
}

// TestDeriveSASMITMSubstitutedSignPub mirrors the x25519 test for the signing
// pubkey; both pubkeys appear in the SAS preimage so both must be validated.
func TestDeriveSASMITMSubstitutedSignPub(t *testing.T) {
	genesis := []byte("genesis-hash-test")
	head := []byte("head-hash-test")
	x25519 := []byte("new-x25519-pub-32bytes-padded-xx")
	legitSign := []byte("legit-sign-pub-65bytes-padded-xx")

	legitimate, err := DeriveSAS(genesis, head, x25519, legitSign)
	if err != nil {
		t.Fatalf("DeriveSAS legitimate: %v", err)
	}

	mitmSign := []byte("mitm--sign-pub-65bytes-padded-xx")
	mitmSAS, err := DeriveSAS(genesis, head, x25519, mitmSign)
	if err != nil {
		t.Fatalf("DeriveSAS MITM sign: %v", err)
	}
	if legitimate == mitmSAS {
		t.Fatalf("MITM failure: substituted sign pubkey produced identical SAS %q", legitimate)
	}
}

// TestDeriveSASEmptyInputsRejected confirms the guard against zero-length hashes.
func TestDeriveSASEmptyInputsRejected(t *testing.T) {
	x25519 := []byte("x25519pub")
	sign := []byte("signpub")

	if _, err := DeriveSAS(nil, []byte("head"), x25519, sign); err == nil {
		t.Fatal("expected error for nil genesis_hash")
	}
	if _, err := DeriveSAS([]byte("genesis"), nil, x25519, sign); err == nil {
		t.Fatal("expected error for nil head_hash")
	}
	if _, err := DeriveSAS(nil, nil, x25519, sign); err == nil {
		t.Fatal("expected error for both nil hashes")
	}
}

// TestDeriveSASKnownVector is the cross-language golden-value test.
// Both sides use the same fixed ASCII byte inputs; the expected SAS was computed
// once and is locked here.  If this test fails the algorithm diverged.
//
// Corresponding TS test: web/src/keys/sas.test.ts "cross-language known vector"
//
// To verify the hardcoded value, run the computation manually:
//   encodeFields("sas/v1", genesis, head, x25519, sign) → SHA-256 → base-36[0:12]
func TestDeriveSASKnownVector(t *testing.T) {
	// Fixed ASCII inputs — stable across platforms and test runs.
	genesis := []byte("test-genesis-hash")
	head := []byte("test-head-hash")
	x25519 := []byte("test-x25519-pub")
	sign := []byte("test-sign-pub")

	const wantSAS = "0004-53td-gr6k"

	got, err := DeriveSAS(genesis, head, x25519, sign)
	if err != nil {
		t.Fatalf("DeriveSAS: %v", err)
	}
	if got != wantSAS {
		t.Fatalf("DeriveSAS known-vector mismatch: got %q, want %q", got, wantSAS)
	}
}

File: web/src/keys/sas.test.ts
/**
 * Tests for SAS (Short Authentication String) derivation ([WM4]).
 *
 * Covers:
 *   - Output format (3 groups of 4 base-36 characters)
 *   - Stability (same inputs → same output)
 *   - [WM4] Named negative: substituted-pubkey MITM attack MUST produce a
 *     different SAS so the human comparison detects the attack.
 *   - Cross-language known vector: deriveSAS matches the Go DeriveSAS output
 *     for fixed ASCII inputs (see internal/secrets/seal/sas_test.go TestDeriveSASKnownVector).
 */

import { describe, it, expect } from "vitest";
import { deriveSAS } from "./sas";

/** SAS format: exactly "xxxx-xxxx-xxxx" where x is a base-36 digit. */
const SAS_FORMAT = /^[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}$/;

// Fixed byte arrays for hermetic unit tests (no real key material needed).
const genesis = new TextEncoder().encode("genesis-hash-test");
const head = new TextEncoder().encode("head-hash-test");
const legitX25519 = new TextEncoder().encode("legit-x25519-pub-32bytes-paddxxx");
const legitSign = new TextEncoder().encode("legit-sign-pub-65bytes-padded-xx");

describe("deriveSAS — format", () => {
  it("returns a 12-character base-36 code in xxxx-xxxx-xxxx format", async () => {
    const sas = await deriveSAS(genesis, head, legitX25519, legitSign);
    expect(sas).toMatch(SAS_FORMAT);
    expect(sas.length).toBe(12 + 2); // 12 chars + 2 hyphens
  });
});

describe("deriveSAS — stability", () => {
  it("produces the same code for identical inputs", async () => {
    const a = await deriveSAS(genesis, head, legitX25519, legitSign);
    const b = await deriveSAS(genesis, head, legitX25519, legitSign);
    expect(a).toBe(b);
  });

  it("produces different codes for different genesis hashes", async () => {
    const a = await deriveSAS(genesis, head, legitX25519, legitSign);
    const b = await deriveSAS(
      new TextEncoder().encode("different-genesis"),
      head,
      legitX25519,
      legitSign,
    );
    expect(a).not.toBe(b);
  });
});

// ── [WM4] Named negative: substituted-pubkey MITM must fail the SAS ──────────

describe("deriveSAS — [WM4] MITM substituted-pubkey failure", () => {
  /**
   * Spec §4 named negative: if an active attacker substitutes the new device's
   * x25519 public key in transit, the approver's independently-derived SAS
   * MUST differ from the enrollee's — so the human comparison catches the
   * attack.  If the SAS were the same for a different pubkey, a MITM could
   * enroll their own key undetected.
   */
  it("[WM4] substituted x25519 pubkey produces a different SAS", async () => {
    const mitmX25519 = new TextEncoder().encode("mitm--x25519-pub-32bytes-paddxxx");

    const legitimate = await deriveSAS(genesis, head, legitX25519, legitSign);
    const mitm = await deriveSAS(genesis, head, mitmX25519, legitSign);

    expect(mitm).not.toBe(legitimate);
  });

  /**
   * Same test for the signing pubkey — both pubkeys appear in the SAS preimage
   * so substituting either one must produce a different code.
   */
  it("[WM4] substituted sign pubkey produces a different SAS", async () => {
    const mitmSign = new TextEncoder().encode("mitm--sign-pub-65bytes-padded-xx");

    const legitimate = await deriveSAS(genesis, head, legitX25519, legitSign);
    const mitm = await deriveSAS(genesis, head, legitX25519, mitmSign);

    expect(mitm).not.toBe(legitimate);
  });
});

// ── Cross-language known vector ───────────────────────────────────────────────

describe("deriveSAS — cross-language known vector", () => {
  /**
   * Fixed ASCII inputs shared with the Go test (TestDeriveSASKnownVector in
   * internal/secrets/seal/sas_test.go).  If this test fails, the TS and Go
   * deriveSAS implementations have diverged.
   */
  it("matches the Go DeriveSAS output for fixed ASCII inputs", async () => {
    const enc = new TextEncoder();
    const sas = await deriveSAS(
      enc.encode("test-genesis-hash"),
      enc.encode("test-head-hash"),
      enc.encode("test-x25519-pub"),
      enc.encode("test-sign-pub"),
    );
    // Hardcoded value verified against Go DeriveSAS in TestDeriveSASKnownVector.
    expect(sas).toBe("0004-53td-gr6k");
  });
});
